### PR TITLE
Ensure __file__ is set when running python scripts dropped onto QGIS, other related tweaks

### PR DESCRIPTION
**Pull request by nyalldawson**
Ensure `__file__` is set when running python scripts dropped onto QGIS, and make sure that the filename is correctly shown when exceptions are raised from dropped scripts or scripts executed through the console (instead of just "").

The first commit should be backported only.

### CODE REVIEW

File: python/console/console_sci.py [hunk 1]
```diff
@@ -437,7 +438,7 @@ def runFile(self, filename):
 
         try:
             # Run the file
-            self.runCommand("exec(Path('{0}').read_text())".format(filename), skipHistory=True)
+            self.runCommand("exec(compile(Path('{0}').read_text(), '{1}', 'exec'))".format(filename, override_file_name or filename), skipHistory=True)
```
Comment (rouault) [inline python/console/console_sci.py:441]: wouldn't there be issues if the filename has a single-quote character ?

Comment (nyalldawson) [reply python/console/console_sci.py:441]: @rouault 

True (although not introduced in this change).

Do you know any reliable inbuilt way to escape this string for use in eval? If we're doing this right then a simple `.replace( "'", "\\'")` won't cut it...

Comment (nyalldawson) [reply python/console/console_sci.py:441]: Looks like elsewhere we use QgsProcessingUtils::stringToPythonLiteral, even outside of processing. I'll do that here.

Comment (nyalldawson) [reply python/console/console_sci.py]: Fixed

**Review by nirvn — APPROVED** [review]

**Pull request merged**